docs/source/configuration.rst
.. # Copyright (c) 2015-2016 Tigera, Inc. All rights reserved.
   #
   #    Licensed under the Apache License, Version 2.0 (the "License"); you may
   #    not use this file except in compliance with the License. You may obtain
   #    a copy of the License at
   #
   #         http://www.apache.org/licenses/LICENSE-2.0
   #
   #    Unless required by applicable law or agreed to in writing, software
   #    distributed under the License is distributed on an "AS IS" BASIS,
   #    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
   #    implied. See the License for the specific language governing
   #    permissions and limitations under the License.

Configuring Calico
==================

This page describes how to configure Calico. We first describe the
configuration of the core Calico component -- Felix --
because this is needed, and configured similarly, regardless of the
surrounding environment (OpenStack, Docker, or whatever). Then,
depending on that surrounding environment, there will be some further
configuration of that environment needed, to tell it to talk to the
Calico components.

Currently we have detailed environment configuration only for OpenStack.
Work on other environments is in progress, and this page will be
extended as that happens.

This page aims to be a complete Calico configuration reference, and
hence to describe all the possible fields, files etc. For a more
task-based approach, when installing Calico with OpenStack on Ubuntu or
Red Hat, please see :doc:`ubuntu-opens-install` or
:doc:`redhat-opens-install`.

System configuration
--------------------

A common problem on Linux systems is running out of space in the conntrack
table, which can cause poor iptables performance. This can happen if you run a
lot of workloads on a given host, or if your workloads create a lot of TCP
connections or bidirectional UDP streams.

To avoid this becoming a problem, we recommend increasing the conntrack table
size. To do so, run the following commands::

    sysctl -w net.netfilter.nf_conntrack_max=1000000
    echo "net.netfilter.nf_conntrack_max=1000000" >> /etc/sysctl.conf

Felix configuration
-------------------

The core Calico component is Felix. (Please see :doc:`architecture`
for the Calico architecture.)

Configuration for Felix is read from one of four possible locations, in order,
as follows.

1. Environment variables.
2. The Felix configuration file.
3. Host specific configuration in etcd.
4. Global configuration in etcd.

The value of any configuration parameter is the value read from the *first*
location containing a value. If not set in any of these locations, most
configuration parameters have defaults, and it should be rare to have to
explicitly set them.

In OpenStack, we recommend putting all configuration into configuration files,
since the etcd database is transient (and may be recreated by the OpenStack
plugin in certain error cases). However, in a Docker environment the use of
environment variables or etcd is often more convenient.

The full list of parameters which can be set is as follows.

+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| Setting                          | Default                               | Meaning                                                                                   |
+==================================+=======================================+===========================================================================================+
| EtcdAddr                         | localhost:4001                        | The location (IP / hostname and port) of the etcd node or proxy that Felix should connect |
|                                  |                                       | to.                                                                                       |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| EtcdScheme                       | http                                  | The protocol type (http or https) of the etcd node or proxy that Felix connects to.       |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| EtcdKeyFile                      | None                                  | The full path to the etcd public key file, as described in :ref:`usingtlswithetcd`        |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| EtcdCertFile                     | None                                  | The full path to the etcd certificate file, as described in :ref:`usingtlswithetcd`       |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| EtcdCaFile                       | "/etc/ssl/certs/ca-certificates.crt"  | The full path to the etcd Certificate Authority certificate file, as described in         |
|                                  |                                       | :ref:`usingtlswithetcd`.  The default value is the standard location of the system trust  |
|                                  |                                       | store.  To disable authentication of the server by Felix, set the value to "none".        |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| DefaultEndpointToHostAction      | DROP                                  | By default Calico blocks traffic from endpoints to the host itself by using an iptables   |
|                                  |                                       | DROP action.  If you want to allow some or all traffic from endpoint to host then set     |
|                                  |                                       | this parameter to "RETURN" (which causes the rest of the iptables INPUT chain to be       |
|                                  |                                       | processed) or "ACCEPT" (which immediately accepts packets).                               |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| DropActionOverride               | DROP                                  | Override for the action taken when a packet would normally be dropped by Calico's         |
|                                  |                                       | firewall rules. This setting is useful when prototyping policy. Note: if the policy is    |
|                                  |                                       | set to 'ACCEPT' or 'LOG-and-ACCEPT'; Calico's security is disabled!                       |
|                                  |                                       |                                                                                           |
|                                  |                                       | In the current implementation, the LOG-and-XXX actions use an iptables LOG action,        |
|                                  |                                       | which logs to syslog.                                                                     |
|                                  |                                       |                                                                                           |
|                                  |                                       | Should be set to one of 'DROP', 'ACCEPT', 'LOG-and-DROP', 'LOG-and-ACCEPT'.               |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| FelixHostname                    | socket.gethostname()                  | The hostname Felix reports to the plugin. Should be used if the hostname Felix            |
|                                  |                                       | autodetects is incorrect or does not match what the plugin will expect.                   |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| MetadataAddr                     | 127.0.0.1                             | The IP address or domain name of the server that can answer VM queries for cloud-init     |
|                                  |                                       | metadata. In OpenStack, this corresponds to the machine running nova-api (or in Ubuntu,   |
|                                  |                                       | nova-api-metadata). A value of 'None' (case insensitive) means that Felix should not set  |
|                                  |                                       | up any NAT rule for the metadata path.                                                    |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| MetadataPort                     | 8775                                  | The port of the metadata server. This, combined with global.MetadataAddr (if not 'None'), |
|                                  |                                       | is used to set up a NAT rule, from 169.254.169.254:80 to MetadataAddr:MetadataPort. In    |
|                                  |                                       | most cases this should not need to be changed.                                            |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| InterfacePrefix                  | cali                                  | The expected prefix for interface names for workload interfaces.  For example, in         |
|                                  |                                       | OpenStack deployments, this should be set to "tap".  Calico polices all traffic to/from   |
|                                  |                                       | interfaces with this prefix.  Calico blocks traffic to/from such interfaces by default.   |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| UsageReportingEnabled            | "true"                                | Periodically report cluster version, hostname, size and guid to projectcalico.org.        |
|                                  |                                       | Receive version deprecation / security warnings.                                          |+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| LogFilePath                      | /var/log/calico/felix.log             | The full path to the felix log. Set to "none" to disable file logging.                    |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| EtcdDriverLogFilePath            | /var/log/calico/felix-etcd.log        | Felix's etcd driver has its own log file. This parameter contains its full path.          |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| LogSeveritySys                   | ERROR                                 | The log severity above which logs are sent to the syslog. Valid values are DEBUG, INFO,   |
|                                  |                                       | WARNING, ERROR and CRITICAL, or NONE for no logging to syslog (all values case            |
|                                  |                                       | insensitive).                                                                             |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| LogSeverityFile                  | INFO                                  | The log severity above which logs are sent to the log file. Valid values as for           |
|                                  |                                       | LogSeveritySys.                                                                           |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| LogSeverityScreen                | ERROR                                 | The log severity above which logs are sent to the stdout. Valid values as for             |
|                                  |                                       | LogSeveritySys.                                                                           |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| StartupCleanupDelay              | 30                                    | Delay, in seconds, before felix does its start-of-day cleanup to remove orphaned iptables |
|                                  |                                       | chains and ipsets.   Before the first cleanup, felix operates in "graceful restart" mode, |
|                                  |                                       | during which it preserves any pre-existing chains and ipsets.                             |
|                                  |                                       |                                                                                           |
|                                  |                                       | In a large deployment you may want to increase this value to give felix more time to      |
|                                  |                                       | load the initial snapshot from etcd before cleaning up.                                   |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| PeriodicResyncInterval           | 3600                                  | Period, in seconds, at which felix does a full resync with etcd and reprograms            |
|                                  |                                       | iptables/ipsets.  Set to 0 to disable periodic resync.                                    |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| IptablesRefreshInterval          | 60                                    | Period, in seconds, at which felix re-applies all iptables state to ensure that no other  |
|                                  |                                       | process has accidentally broken Calico's rules.  Set to 0 to disable iptables refresh.    |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| MaxIpsetSize                     | 1048576                               | Maximum size for the ipsets used by Felix to implement tags.  Should be set to a number   |
|                                  |                                       | that is greater than the maximum number of IP addresses that are ever expected in a tag.  |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| IptablesMarkMask                 | 0xff000000                            | Mask that Felix selects its IPTables Mark bits from.  Should be a 32 bit hexadecimal      |
|                                  |                                       | number with at least 8 bits set, none of which clash with any other mark bits in use on   |
|                                  |                                       | the system.                                                                               |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| PrometheusMetricsEnabled         | "false"                               | Set to "true" to enable the experimental Prometheus metrics server in Felix.              |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| PrometheusMetricsPort            | 9091                                  | TCP port that the Prometheus metrics server should bind to.                               |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| EtcdDriverPrometheusMetricsPort  | 9092                                  | TCP port that the Prometheus metrics server in the etcd driver process should bind to.    |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+
| FailsafeInboundHostPorts         | 22                                    | Comma-delimited list of TCP ports that Felix will allow incoming traffic to host          |
|                                  |                                       | endpoints on irrespective of the security policy.  This is useful to avoid accidently     |
|                                  |                                       | cutting off a host with incorrect configuration.  The default value allows ssh access.    |
+----------------------------------+---------------------------------------+-------------------------------------------------------------------------------------------+

CHANGES.md
# Changelog

Pull request test!  Do not merge!  OK?

## 1.1.0

- Improve the documentation about upgrading a Calico/OpenStack system.
- Fix compatibility with latest OpenStack code (oslo_config).
- Use posix_spawn to improve Felix's performance under heavy load.
- Explicitly use and enable the kernel's reverse path filtering function,
  and remove our iptables anti-spoofing rules, which were not as robust.

## 1.0.0

- Add support for setting MTU on IP-in-IP device.
- Enhance BIRD configuration and documentation for graceful restart.

## 0.28

- Felix now restarts if its etcd configuration changes.
- Felix now periodically refreshes iptables to be robust to other processes
  corrupting its chains.
- More thorough resynchronization of etcd from the Neutron mechanism driver.
- Added process-specific information to the diagnostics dumps from Felix.

## 0.27

- Limit number of concurrent shell-outs in felix to prevent file descriptor
  exhaustion.
- Have felix periodically resync from etcd and force-refresh the dataplane.
- Stop restarting Felix on Ubuntu if it fails more than 5 times in 10 seconds.
- Move DHCP checksum calculation to Neutron.
- Get all fixed IPs for a port.

## 0.26

- Update and improve security model documentation.
- Streamline conntrack rules, move them to top-level chains to avoid
  duplication.
- Narrow focus of input iptables chain so that it only applies to
  Calico-handled traffic.
- Provide warning log when attempting to use Neutron networks that are not of
  type 'local' or 'flat' with Calico.
- Handle invalid JSON in IPAM key in etcd.
- Move all log rotation into logrotate and out of Felix, to prevent conflicts.
- Change log rotation strategy for logrotate to not rotate small log files.
- Delay starting the Neutron resynchronization thread until after all the
  necessary state has been configured, to avoid race conditions.
- Prevent systemd restarting Felix when it is killed by administrators.

## 0.25

- Remove stale conntrack entries when an endpoint's IP is removed.
- #672: Fix bug where profile chain was left empty instead of being
  stubbed out.
- Improve security between endpoint and host and simplify INPUT chain logic.

## 0.24

- Add Felix statistics logging on USR1 signal.
- Add support for routing over IP-in-IP interfaces in order to make it
  easier to evaluate Calico without reconfiguring underlying network.
- Reduce felix occupancy by replacing endpoint dictionaries by "struct"
  objects.
- Allow different hosts to have different interface prefixes for combined
  OpenStack and Docker systems.
- Add missing support for 0 as a TCP port.
- Add support for arbitrary IP protocols.
- Intern various IDs in felix to reduce occupancy.
- Fix bug where Calico may not propagate security group rule changes from
  OpenStack.
- Reduced logspam from Calico Mechanism Driver.

## 0.23

- Reset ARP configuration when endpoint MAC changes.
- Forget about profiles when they are deleted.
- Treat bad JSON as missing data.
- Add instructions for Kilo on RHEL7.
- Extend diagnostics script to collect etcd and RabbitMQ information.
- Improve BIRD config to prevent NETLINK: File Exists log spam.
- Reduce Felix logging volume.

## 0.22

- Updated Mechanism driver to specify fixed MAC address for Calico tap
  interfaces.
- Prevent the possibility of gevent context-switching during garbage collection
  in Felix.
- Increase the number of file descriptors available to Felix.
- Firewall input characters in profiles and tags.
- Implement tree-based dispatch chains to improve IPTables performance with
  many local endpoints.
- Neutron mechanism driver patches and docs for OpenStack Kilo release.
- Correct IPv6 documentation for Juno and Kilo.

## 0.21

- Support for running multiple neutron-server instances in OpenStack.
- Support for running neutron-server API workers in OpenStack.
- Calico Mechanism Driver now performs leader election to control state
  resynchronization.
- Extended data model to support multiple security profiles per endpoint.
- Calico Mechanism Driver now attempts to delete empty etcd directories.
- Felix no longer leaks memory when etcd directories it watches are deleted.
- Fix error on port creation where the Mechanism Driver would create, delete,
  and then recreate the port in etcd.
- Handle EtcdKeyNotFound from atomic delete methods
- Handle etcd cluster ID changes on API actions
- Fix ipsets cleanup to correctly iterate through stopping ipsets
- Ensure that metadata is not blocked by over-restrictive rules on outbound
  traffic
- Updates and clarifications to documentation


go/felix/config/config.go
// Copyright (c) 2016 Tigera, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package config

import (
	"errors"
	"fmt"
	"github.com/golang/glog"
	"github.com/tigera/libcalico-go/lib/api"
	"github.com/tigera/libcalico-go/lib/backend/etcd"
	"net"
	"os"
	"reflect"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var (
	IfaceNameRegexp = regexp.MustCompile(`^[a-zA-Z0-9_-]{1,15}$`)
	AuthorityRegexp = regexp.MustCompile(`^[^:/]+:\d+$`)
	HostnameRegexp  = regexp.MustCompile(`^[a-zA-Z0-9_.-]+$`)
)

const (
	maxUint = ^uint(0)
	maxInt  = int(maxUint >> 1)
	minInt  = -maxInt - 1
)

// Source of a config value.  Values from higher-numbered sources override
// those from lower-numbered sources.  Note: some parameters (such as those
// needed to connect to the datastore) can only be set from a local source.
type Source uint8

const (
	Default = iota
	DatastoreGlobal
	DatastorePerHost
	ConfigFile
	EnvironmentVariable
)

func (source Source) String() string {
	switch source {
	case Default:
		return "<default>"
	case DatastoreGlobal:
		return "datastore (global)"
	case DatastorePerHost:
		return "datastore (per-host)"
	case ConfigFile:
		return "config file"
	case EnvironmentVariable:
		return "environment variable"
	}
	return fmt.Sprintf("<unknown(%v)>", uint8(source))
}

func (source Source) Local() bool {
	switch source {
	case Default, ConfigFile, EnvironmentVariable:
		return true
	default:
		return false
	}
}

// Config contains the best, parsed config values loaded from the various sources.
// We use tags to control the parsing and validation.
type Config struct {
	// Configuration parameters.

	FelixHostname string `config:"hostname;;local,non-zero"`

	EtcdAddr      string   `config:"authority;127.0.0.1:2379;local"`
	EtcdScheme    string   `config:"oneof(http,https);http;local"`
	EtcdKeyFile   string   `config:"file(must-exist);;local"`
	EtcdCertFile  string   `config:"file(must-exist);;local"`
	EtcdCaFile    string   `config:"file(must-exist);;local"`
	EtcdEndpoints []string `config:"endpoint-list;;local"`

	StartupCleanupDelay       int `config:"int;30"`
	PeriodicResyncInterval    int `config:"int;3600"`
	HostInterfacePollInterval int `config:"int;10"`

	IptablesRefreshInterval int `config:"int;60"`

	MetadataAddr string `config:"hostname;127.0.0.1;die-on-fail"`
	MetadataPort int    `config:"int(0,65535);8775;die-on-fail"`

	InterfacePrefix string `config:"iface-name;cali;non-zero,die-on-fail"`

	DefaultEndpointToHostAction string `config:"oneof(DROP,RETURN,ACCEPT);DROP;non-zero,die-on-fail"`
	DropActionOverride          string `config:"oneof(DROP,ACCEPT,LOG-and-DROP,LOG-and-ACCEPT);DROP;non-zero,die-on-fail"`

	LogFilePath           string `config:"file;/var/log/calico/felix.log;die-on-fail"`
	EtcdDriverLogFilePath string `config:"file;/var/log/calico/felix-etcd.log"`

	LogSeverityFile   string `config:"oneof(DEBUG,INFO,WARNING,ERROR,CRITICAL);INFO"`
	LogSeveritySys    string `config:"oneof(DEBUG,INFO,WARNING,ERROR,CRITICAL);INFO"`
	LogSeverityScreen string `config:"oneof(DEBUG,INFO,WARNING,ERROR,CRITICAL);INFO"`

	IpInIpEnabled    bool   `config:"bool;false"`
	IpInIpMtu        int    `config:"int;1440;non-zero"`
	IpInIpTunnelAddr net.IP `config:"ipv4;"`

	ReportingIntervalSecs int `config:"int;30"`
	ReportingTTLSecs      int `config:"int;90"`

	EndpointReportingEnabled   bool    `config:"bool;false"`
	EndpointReportingDelaySecs float64 `config:"float;1.0"`

	MaxIpsetSize int `config:"int;1048576;non-zero"`

	IptablesMarkMask uint32 `config:"mark-bitmask;0xff000000;non-zero,die-on-fail"`

	PrometheusMetricsEnabled        bool `config:"bool;false"`
	PrometheusMetricsPort           int  `config:"int(0,65535);9091"`
	EtcdDriverPrometheusMetricsPort int  `config:"int(0,65535);9092"`

	FailsafeInboundHostPorts  []int `config:"port-list;22;die-on-fail"`
	FailsafeOutboundHostPorts []int `config:"port-list;2379,2380,4001,7001;die-on-fail"`

	// State tracking.

	// nameToSource tracks where we loaded each config param from.
	nameToSource map[string]Source
	RawValues    map[string]string
	Err          error
}

// Load parses and merges the rawData from one particular source into this config object.
// If there is a config value already loaded from a higher-priority source, then
// the new value will be ignored (after validation).
func (config *Config) UpdateFrom(rawData map[string]string, source Source) (changed bool, err error) {
	glog.V(2).Infof("Merging in config from %v: %v", source, rawData)
	for rawName, rawValue := range rawData {
		currentSource := config.nameToSource[rawName]
		param, ok := knownParams[strings.ToLower(rawName)]
		if !ok {
			if source >= currentSource {
				// Stash the raw value in case it's useful for
				// a plugin.  Since we don't know the canonical
				// name, use the raw name.
				config.RawValues[rawName] = rawValue
				config.nameToSource[rawName] = source
			}
			glog.Warningf("Ignoring unknown configuration parameter: %v", rawName)
			continue
		}
		metadata := param.getMetadata()
		name := metadata.Name
		if metadata.Local && !source.Local() {
			glog.Warningf("Ignoring local-only configuration for %v from %v",
				name, source)
			continue
		}

		glog.V(2).Infof("Parsing value for %v: %v (from %v)",
			name, rawValue, source)
		var value interface{}
		if strings.ToLower(rawValue) == "none" {
			if metadata.NonZero {
				err = errors.New("Non-zero field cannot be set to none")
				glog.Errorf(
					"Failed to parse value for %v: %v from source %v. %v",
					name, rawValue, source, err)
				config.Err = err
				return
			}
			glog.V(2).Infof("Value set to 'none', replacing with zero-value: %#v.",
				value)
			value = metadata.ZeroValue
		} else {
			value, err = param.Parse(rawValue)
			if err != nil {
				glog.Errorf("%v (source %v)", err, source)
				if metadata.DieOnParseFailure {
					glog.Errorf("Cannot continue with invalid value for %v.", name)
					config.Err = err
					return
				} else {
					glog.Errorf("Replacing invalid value with default value for %v: %v",
						name, metadata.Default)
					value = metadata.Default
					err = nil
				}
			}
		}

		field := reflect.ValueOf(config).Elem().FieldByName(name)
		currentValue := field.Interface()
		if currentValue == value {
			glog.V(3).Infof("Value of %v hasn't changed, skipping.", name)
			continue
		}

		glog.V(2).Infof("Parsed value for %v: %v (from %v)",
			name, value, source)
		if source < currentSource {
			glog.V(2).Infof("Skipping config value for %v from %v; "+
				"already have a value from %v", source, currentSource)
			continue
		}
		glog.V(2).Infof(
			"Now using %v value from %v (preferred to "+
				"previous value from %v)",
			name, source, currentSource)
		field.Set(reflect.ValueOf(value))
		if config.RawValues[name] != rawValue {
			glog.V(1).Infof("Configuration value %v changed from %v to %v",
				name, config.RawValues[name], rawValue)
			changed = true
		}
		config.RawValues[name] = rawValue
		config.nameToSource[name] = source
	}
	return
}

func (config *Config) EndpointReportingDelay() time.Duration {
	return time.Duration(config.EndpointReportingDelaySecs*1000000) * time.Microsecond
}

func (config *Config) DatastoreConfig() api.ClientConfig {
	var etcdEndpoints string
	if len(config.EtcdEndpoints) == 0 {
		etcdEndpoints = config.EtcdScheme + "://" + config.EtcdAddr
	} else {
		etcdEndpoints = strings.Join(config.EtcdEndpoints, ",")
	}
	etcdCfg := &etcd.EtcdConfig{
		EtcdEndpoints:  etcdEndpoints,
		EtcdKeyFile:    config.EtcdKeyFile,
		EtcdCertFile:   config.EtcdCertFile,
		EtcdCACertFile: config.EtcdCaFile,
	}
	return api.ClientConfig{
		BackendType:   api.EtcdV2,
		BackendConfig: etcdCfg,
	}
}

// Validate() performs cross-field validation.
func (config *Config) Validate() (err error) {
	if config.FelixHostname == "" {
		err = errors.New("Failed to determine hostname")
	}

	if len(config.EtcdEndpoints) == 0 {
		if config.EtcdScheme == "" {
			err = errors.New("EtcdEndpoints and EtcdScheme both missing")
		}
		if config.EtcdAddr == "" {
			err = errors.New("EtcdEndpoints and EtcdAddr both missing")
		}
	}

	if err != nil {
		config.Err = err
	}
	return
}

var knownParams map[string]param

func loadParams() {
	knownParams = make(map[string]param)
	config := Config{}
	kind := reflect.TypeOf(config)
	metaRegexp := regexp.MustCompile(`^([^;(]+)(?:\(([^)]*)\))?;` +
		`([^;]*)(?:;` +
		`([^;]*))?$`)
	for ii := 0; ii < kind.NumField(); ii++ {
		field := kind.Field(ii)
		tag := field.Tag.Get("config")
		if tag == "" {
			continue
		}
		captures := metaRegexp.FindStringSubmatch(tag)
		if len(captures) == 0 {
			glog.Fatalf("Failed to parse metadata for config param %v", field.Name)
		}
		glog.V(3).Infof("%v: metadata captures: %#v", field.Name, captures)
		kind := captures[1]       // Type: "int|oneof|bool|port-list|..."
		kindParams := captures[2] // Parameters for the type: e.g. for oneof "http,https"
		defaultStr := captures[3] // Default value e.g "1.0"
		flags := captures[4]
		var param param
		var err error
		switch kind {
		case "bool":
			param = &boolParam{}
		case "int":
			min := minInt
			max := maxInt
			if kindParams != "" {
				minAndMax := strings.Split(kindParams, ",")
				min, err = strconv.Atoi(minAndMax[0])
				if err != nil {
					glog.Fatalf("Failed to parse min value for %v", field.Name)
				}
				max, err = strconv.Atoi(minAndMax[1])
				if err != nil {
					glog.Fatalf("Failed to parse max value for %v", field.Name)
				}
			}
			param = &intParam{Min: min, Max: max}
		case "int32":
			param = &int32Param{}
		case "mark-bitmask":
			param = &markBitmaskParam{}
		case "float":
			param = &floatParam{}
		case "iface-name":
			param = &regexpParam{Regexp: IfaceNameRegexp,
				Msg: "invalid Linux interface name"}
		case "file":
			if kindParams != "" && kindParams != "must-exist" {
				glog.Fatalf("Bad type params for 'file': %#v",
					kindParams)
			}
			param = &fileParam{MustExist: kindParams == "must-exist"}
		case "authority":
			param = &regexpParam{Regexp: AuthorityRegexp,
				Msg: "invalid URL authority"}
		case "ipv4":
			param = &ipv4Param{}
		case "endpoint-list":
			param = &endpointListParam{}
		case "port-list":
			param = &portListParam{}
		case "hostname":
			param = &regexpParam{Regexp: HostnameRegexp,
				Msg: "invalid hostname"}
		case "oneof":
			options := strings.Split(kindParams, ",")
			lowerCaseToCanon := make(map[string]string)
			for _, option := range options {
				lowerCaseToCanon[strings.ToLower(option)] = option
			}
			param = &oneofListParam{
				lowerCaseOptionsToCanonical: lowerCaseToCanon}
		default:
			glog.Fatalf("Unknown type of parameter: %v", kind)
		}

		metadata := param.getMetadata()
		metadata.Name = field.Name
		metadata.ZeroValue = reflect.ValueOf(config).FieldByName(field.Name).Interface()
		if strings.Index(flags, "non-zero") > -1 {
			metadata.NonZero = true
		}
		if strings.Index(flags, "die-on-fail") > -1 {
			metadata.DieOnParseFailure = true
		}
		if strings.Index(flags, "local") > -1 {
			metadata.Local = true
		}

		if defaultStr != "" {
			// Parse the default value and save it in the metadata. Doing
			// that here ensures that we syntax-check the defaults now.
			defaultVal, err := param.Parse(defaultStr)
			if err != nil {
				glog.Fatalf("Invalid default value: %v", err)
			}
			metadata.Default = defaultVal
		} else {
			metadata.Default = metadata.ZeroValue
		}
		knownParams[strings.ToLower(field.Name)] = param
	}
}

func New() *Config {
	if knownParams == nil {
		loadParams()
	}
	p := &Config{
		RawValues:    make(map[string]string),
		nameToSource: make(map[string]Source),
	}
	for _, param := range knownParams {
		param.setDefault(p)
	}
	hostname, err := os.Hostname()
	if err != nil {
		glog.Warningf("Failed to get hostname from kernel, "+
			"trying HOSTNAME variable: %v", err)
		hostname = os.Getenv("HOSTNAME")
	}
	p.FelixHostname = hostname
	return p
}

type param interface {
	getMetadata() *metadata
	Parse(raw string) (result interface{}, err error)
	setDefault(*Config)
}
